src/cpp/session/modules/SessionProjectTemplate.cpp

```diff
@@ -0,0 +1,384 @@
+
+#define kProjectTemplateLocal "(local)"
+#define kRStudioProjectTemplatesPath "rstudio/templates"
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace projects {
+namespace templates {
+
+namespace {
+
+class ProjectTemplateRegistry
+{
+public:
+   
+   void add(const std::string& pkgName, const ProjectTemplateDescription& description)
+   {
+      registry_[pkgName].push_back(description);
+   }
+   
+   std::vector<ProjectTemplateDescription>& get(const std::string& pkgName)
+   {
+      return registry_[pkgName];
+   }
+   
+   std::size_t size()
+   {
+      return registry_.size();
+   }
+   
+   json::Value toJson()
+   {
+      json::Object object;
+      
+      BOOST_FOREACH(const std::string& pkgName, registry_ | boost::adaptors::map_keys)
+      {
+         json::Array array;
+         BOOST_FOREACH(const ProjectTemplateDescription& description, registry_[pkgName])
+         {
+            array.push_back(description.toJson());
+         }
+         object[pkgName] = array;
+      }
+      
+      return object;
+   }
+
+   std::map<
+      std::string,
+      std::vector<ProjectTemplateDescription>
+   > registry_;
+};
+
+ProjectTemplateRegistry& projectTemplateRegistry()
+{
+   static ProjectTemplateRegistry instance;
+   return instance;
+}
+
+class ProjectTemplateIndexer : public ppe::Indexer
+{
+public:
+   
+   explicit ProjectTemplateIndexer(const std::string& resourcePath)
+      : ppe::Indexer(resourcePath)
+   {
+   }
+   
+   void addIndexingFinishedCallback(boost::function<void()> callback)
+   {
+      callbacks_.push_back(callback);
+   }
+   
+   Error parseResourceFile(const std::string& pkgName,
+                           const FilePath& resourcePath,
+                           ProjectTemplateDescription* pDescription)
+   {
+      Error error;
+      
+      // read dcf contents
+      std::string contents;
+      error = core::readStringFromFile(resourcePath, &contents, string_utils::LineEndingPosix);
+      if (error)
+         return error;
+      
+      // attempt to parse as DCF -- multiple newlines used to separate records
+      boost::regex reSeparator("\\n{2,}");
+      boost::sregex_token_iterator it(contents.begin(), contents.end(), reSeparator, -1);
+      boost::sregex_token_iterator end;
+      
+      for (; it != end; ++it) {
+         // invoke parser on current record
+         std::map<std::string, std::string> fields;
+         std::string errorMessage;
+         error = text::parseDcfFile(*it, true, &fields, &errorMessage);
+         if (error)
+            continue;
+         
+         // populate project template description based on fields
+         error = ProjectTemplateDescription::populate(resourcePath, fields, pDescription);
+         if (error)
+            continue;
+      }
+      
+      return error;
+   }
+   
+private:
+   void onIndexingStarted()
+   {
+      pRegistry_ = boost::make_shared<ProjectTemplateRegistry>();
```

Comment:
It looks like this shared pointer manages the lifetime of the global`ProjectTemplateRegistry`. If that's the case then does resetting it cause the pointer to release the registry, so that there is briefly no registry until indexing completes?